### PR TITLE
Make social icon navigation one arrow keypress

### LINKED ISSUE #64937: Broken Keyboard Interaction for moving between and using social icons
### Description

- Moving between each social icon is two keypresses (one to get to the `<li>` with the blockProps and one to the nested `<button>`).
- The `<li>` does not have an onClick to open the popover but the <button> does, meaning you can only press enter to open the popover when on the `button`, but you can't tell which one you're focused on.

### Step-by-step reproduction instructions

- Add a social icons block
- Using the keyboard, move focus to the social icons
- Press Enter
- Nothing will happen, as you're on the LI
- Press the down or right arrows
- Focus should be on the same social icon, but have shifted slightly
- Press Enter
- The popover should open

This should only be one keypress to move between them, and when focus is on the element, enter should always open the popover.

### Screenshots, screen recording, code snippet

https://github.com/user-attachments/assets/d53843f3-6b4a-4827-a3a8-80d9a475f810

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [X] Yes

### PR DESCRIPTION
**Pull request by jeryj**
Fixes https://github.com/WordPress/gutenberg/issues/64937

## What?

Moves the `blockProps` for the social icon from the `<li>` to the `<button>`.
## Why?

1. Moving between each social icon was two keypresses (one to get to the `<li>` and one to the nested `<button>`.
2. The `<li>` did not have an onClick to open the popover but the `<button>` did, meaning you can only press enter to open the popover when on the `<button> but you can't tell which one you're focused on. 

## How?

By moving the `blockProps` for the social icon from the `<li>` to the `<button>`, we get the appropriate behavior of being able to select and create social icons with a clean keyboard behavior. The caveat is to preserve the CSS, we needed to add the `.wp-block-social-link` to both the `li` and the `button`. This allows us the least amount of changes to the CSS, preventing a lot of potential CSS conflicts and styling issues for sites.

## Testing Instructions

Check for visual regressions on both the frontend and editor

### Testing Instructions for Keyboard

- Add a social icons block
- Using the keyboard, move focus to the social icons
- When on a social icon, moving left/right between them should only be one keypress
- Press Enter on a social icon
- Focus should be within the popover
- Submit a value
- It should be saved and focus moved to the social icon button
- Press Enter
- Focus should be within the popover
- Press Escape
- The popover should close and focus moved to the social icon button

## Screenshots or screencast 

https://github.com/user-attachments/assets/63c78276-db23-4881-b0c5-861ab7b532e6

### CODE REVIEW

**Comment by alexstine** [comment]
@jeryj Make sure this doesn't break other interactions such as multiselect. I tried to globally remove this `tabindex` from `useBlockProps` and ended up breaking a bunch of different focus scenarios. One such example is the ability to focus the block wrapper to delete or insert a block via backspace or enter. I do not have bandwidth to help further but thought you should know I've already been down this path.

**Comment by afercia** [comment]
@jeryj Thanks for this PR. Quickly tested and it looks like a good first step. Some considerations:

The most important point is that arrows navigation wouldn't work well for Windows screen readers unless we make sure the screen readere switches to 'focus mode'. Otherwise, arrows are reserved for internal usage in screen readers 'browse more'. IN other cases, the editor uses an ARIA role to do this for example the `application` role in the list view. In this case, I wouldn't know what the best option is. @alexstine any thoughts?

Secondly: while it makes totally sense to use a list on the front end, why we're using a list in the editor? It doesn't make much sense to use a list and then override its semantics by using `role="document"`. The markup in the editor is soemthign liek the following:

```
ul 
    role="document"

li
    role="document"
    aria-label="Block: GitHub"
    tabindex="0" becomes tabindex="-1" in this PR

button 
    labeled by its content e.g. 'GitHub'
```

The `li` element role is actuall `document`. Still, it's labeled, which is incorrect.
Why the `li` element needs a tabindex attribute in the first place?
Should we consider to entirely remove the list in the editor and only use buttons?

**Review by afercia — COMMENTED** [review]
Please see my last comment

**Review by getdave — COMMENTED** [review]
To me this feels like an improvement over what we have currently (i.e. two key presses to navigate between link items).

The only caveat is if this breaks multi-select like @alexstine suggested. I wasn't sure how best to verify that one. I was trying SHIFT + ARROWs to select...

**Comment by alexstine** [comment]
@afercia The only solution here is using role application for the editor canvas but it's a huge lift and requires tons of testing. As I've stated in the passed, role document is my fault because I attempted to fix a legacy JAWS bug. That was a lesson learned on my part. We used role group before but even that was bad and would not fix this issue. Removing the role all together from `useBlockProps` is an interesting experience because it restores the semantics of all blocks. For example, inserting a table block with no role document or group will now make the block discoverable as a table.

Someone will have to take this on one of these days and then fix the endless E2E tests. Furthermore, we'd need real user testing of every block and situation to ensure we're not making the editor worse than it currently is.

I do not think we can remove the ability for the block wrapper to gain focus though, this totally removes the ability to easily delete or insert a block after the current position. Admittedly, it is a little known feature to keyboard users.

**Comment by jeryj** [comment]
Thanks for the catch, @alexstine! I was assuming this fix was too good to be true :) 

> I do not think we can remove the ability for the block wrapper to gain focus though

Yeah, you're right. Thanks for catching this before I went too far down the path!

> while it makes totally sense to use a list on the front end, why we're using a list in the editor?

@afercia I believe to make sure CSS styles match so that the editor is an accurate representation of the frontend site.

Some potential solutions:
- Make the `<button>` the block wrapper instead of the `<li>`. This means passing the blockProps to the `button`, which I did try at first but it messed up all the styling since the `<li>` has some classNames on it. That's going to take some interesting reworking. 
- Another hackier but potential solution might be to make the `<li>` a `role="button"` and the `button` a `role="none"`. Is this an especially bad idea or could it work here?

I'll play around with them and see if there's something we can do to make this better.

**Comment by alexstine** [comment]
@jeryj What if you just add the block props to button and adjust the CSS styling? Although, I am wondering why `useBlockProps` I think use-first-focus-element is not working? When a block is selected, it should focus the button inside the li because the li is not a form element.

**Comment by afercia** [comment]
> Another hackier but potential solution might be to make the <li> a role="button" and the button a role="none". Is this an especially bad idea or could it work here?

Technically, it's possible to make any element behave like a button but that would mean replicating all the semantics and interaction of a native button, including making sure the button can be acitvated on spacebar _keyup_. It's so overkill that I wouldn't recommend it.

**Comment by jeryj** [comment]
I ended up moving the `..blockProps` to the `<button />`. Seems to be working well! Ready for testing 👍🏻

**Comment by afercia** [comment]
To my understanding, by moving the props from the `li` to the `button`, also some of the styles are moved. This impacts the border radius of the button and the focus style, which is now different. See screenshot below, before and after.

before

**Comment by jeryj** [comment]
> This impacts the border radius of the button and the focus style, which is now different. 

@afercia My suspicion from the code is that the `<li>` with `blockProps` had the default block focus style, then because the button also had focus it was given a distinct focus style to distinguish it from the `<li>`. Now that the `blockProps` are on the button, I think it's best that the focus style matches the default block selection focus style (thinner border, no border radius). So, I think the focus style changing to the default block focus style is the right decision.

**Comment by afercia** [comment]
@jeryj thanks for your feedback. I have no strong opinions on whether to consider the social icons in the editor 'blocks' or 'buttons', the most important thing is that there is a focus style whatever its shape is 🙂

**Comment by getdave** [comment]
I rebased here because my recent changes to the UX in Core weren't reflected but it seems I made a mistake and reverted @jeryj's progress. He will have his local copy which he can use to update this when he's back. My apologies - it was early in the morning 🤦

**Comment by jeryj** [comment]
It happens! Rebased and force pushed :)

File: packages/block-library/src/social-link/edit.js [hunk 1]
```diff
@@ -108,12 +109,19 @@ const SocialLinkEdit = ( {
...
-		'wp-social-link__is-incomplete': ! url,
-		[ `has-${ iconColor }-color` ]: iconColor,
-		[ `has-${ iconBackgroundColor }-background-color` ]:
-			iconBackgroundColor,
-	} );
+	const wrapperClasses = clsx(
+		'wp-social-link',
+		// Manually adding this class for backwards compatibility of CSS when moving the
+		// blockProps from the li to the button: https://github.com/WordPress/gutenberg/pull/64883
+		'wp-block-social-link',
```
Comment (jeryj) [inline packages/block-library/src/social-link/edit.js:120]: `wp-block-social-link` is the only one that was added, and although it's technically "wrong" as only the social link block should get this class, by adding it we greatly reduce the chance of breaking user's CSS and this class is also output on the LI on the frontend.

File: packages/block-library/src/social-link/edit.js [hunk 2]
```diff
@@ -165,13 +178,14 @@ const SocialLinkEdit = ( {
...
-					ref={ setPopoverAnchor }
-					onClick={ () => setPopover( true ) }
-					aria-haspopup="dialog"
-				>
+			<li
+				className={ wrapperClasses }
+				style={ {
+					color: iconColorValue,
+					backgroundColor: iconBackgroundColorValue,
+				} }
```
Comment (jeryj) [inline packages/block-library/src/social-link/edit.js:196]: These were all added via the blockProps before, so we're adding them manually to preserve their styles.

File: packages/block-library/src/social-link/edit.js [hunk 3]
```diff
@@ -127,11 +135,16 @@ const SocialLinkEdit = ( {
...
-			color: iconColorValue,
-			backgroundColor: iconBackgroundColorValue,
+		className: 'wp-block-social-link-anchor',
+		ref: useMergeRefs( [ setPopoverAnchor, ref ] ),
+		onClick: () => setPopover( true ),
+		onKeyDown: ( event ) => {
+			if ( event.keyCode === ENTER ) {
+				event.preventDefault();
+				setPopover( true );
+			}
```
Comment (jeryj) [inline packages/block-library/src/social-link/edit.js:151]: It needs the `onKeyDown`, as otherwise it gets caught by [an event listener in useBlockProps for adding a new block on enter keypress](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/block-list/use-block-props/use-selected-block-event-handlers.js#L60). We're not removing any functionality though, as enter on keypress to add a new icon did not work on trunk either.

File: packages/block-library/src/social-link/editor.scss [hunk 4]
```diff
@@ -21,13 +21,10 @@
...
-	&:focus:not(:disabled) {
-		border-radius: 2px;
-		box-shadow: 0 0 0 var(--wp-admin-border-width-focus) var(--wp-admin-theme-color);
-
-		// Windows High Contrast mode will show this outline, but not the box-shadow.
-		outline: 3px solid transparent;
+	// Override the shared `.wp-block-social-link` class used on both the li and button
+	// due to backwards compatibility from moving the blockProps from the li to the button.
+	&:hover {
+		transform: none;
```
Comment (jeryj) [inline packages/block-library/src/social-link/editor.scss:27]: This only impacts the editor style. This transform is already applied via the LI, so all we're doing is preventing that transform from triggering twice.

**Comment by afercia** [comment]
@jeryj @getdave thanks for working on this. What's left here to get this merged?
I'd love to fix also https://github.com/WordPress/gutenberg/issues/64972 but it would be good to merge this first. Thanks.

**Comment by getdave** [comment]
> What's left here to get this merged?

More reviews and testing I believe.

**Review by getdave — COMMENTED** [review]
In my testing I was able to move to the block and use the arrow keys and `ENTER` to perform all the key interactions. I was even able to move to the `+` inserter using arrows to add new blocks.

Overall this seems like an improvement.

**Comment by jeryj** [comment]
Anyone willing to give it an approving review?

**Comment by afercia** [comment]
To my understanding, the markup on the front end doesn't change and that's great.
In the editor, it does change and testing with Safari and VoiceOver, there may be a problem. B moving the props to the button, the button gets also a `role=document` which is normally set to the block (the li element) and it's now moved to the button. Simplified markup in the editor:  

Before:

```
<li tabindex="0" class="block-editor-block-list__block wp-block wp-social-link wp-social-link-github wp-block-social-link" id="block-39ba69a7-6f19-478e-9a29-ef99aeb041f0" role="document" aria-label="Block: GitHub" data-block="39ba69a7-6f19-478e-9a29-ef99aeb041f0" data-type="core/social-link" data-title="GitHub">
    <button class="wp-block-social-link-anchor" aria-haspopup="dialog"><span class="wp-block-social-link-label screen-reader-text">GitHub</span></button>
</li>
```

After:

```
<li class="wp-social-link wp-block-social-link wp-social-link__list-item wp-social-link-github">
    <button aria-haspopup="dialog" tabindex="0" class="block-editor-block-list__block wp-block wp-block-social-link-anchor wp-block-social-link" id="block-320dc936-1769-4850-8a98-bb907a9a2ec3" role="document" aria-label="Block: GitHub" data-block="320dc936-1769-4850-8a98-bb907a9a2ec3" data-type="core/social-link" data-title="GitHub"><span class="wp-block-social-link-label screen-reader-text">GitHub</span></button>
</li>
```

Since the button native role is now overridden by `role="document"`, the button is announced as 'document'. VoiceOver dosn't even announce `aria-haspopup="dialog"` because `aria-haspopup` is unexpected on a document.

For screen reader users this makes ery difficult to understand how to interact with the social icons nested blocks and at this point I'm not sure moving _all_ the props to the button is the best option. A few screenshots to better illustrate:

Before:

The LI element is the block and it's announced as such:

Screenshot 2024-09-25 at 09 23 08

Navigating with Control+Option+Right arrow to the button, the button is announced as a button with a dialog pop-up, which is what I would expect:

Screenshot 2024-09-25 at 09 23 15

After:

The LI element is still announced as block, probably because it now _contains_ the block:

Screenshot 2024-09-25 at 09 25 50

Navigating with Control+Option+Right arrow to the button, the button is not announced as button any longer. The only thin that is announced is the aria-label `aria-label="Block: GitHub"`. No information that is'a an interactive button that opens a dialog.

Screenshot 2024-09-25 at 09 27 00

Basically, while moving the props to the button solves the keyboard interaction issue, it removes any information on how to interact with the button, which is a problem for screen reader users.

The root problem is that each `LI` is actually treateed as a nested block. As such it's focusable and labeled. Would it make sense to keep the props on the LI and only remove the tabindex?

**Comment by getdave** [comment]
I guess another option is to remove the `ul > li` markup entirely _in the editor_ (only) and just use `div > buttons`. I know that we've established that markup between editor and front end doesn't _have_ to match.

**Comment by afercia** [comment]
Couple thoughts:

1
One option could be removing `tabIndex` from the blockProps. This way the `LI`, which is the actual inner block, wouldn't be focusable any longer and focus would go to the button. However, this would be an exception that is hard to justify as all blocks are supposed to be focusable.

2
What is the difference about focus management with similar blocks? For example the Buttons block:
- The parent block is focusable, that's not an issue.
- The inner Button block is a `div` element with `tabindex=0` so it's focusable.
- It contains a contenteditable element (RichText) that is focusable.

However, when tabbing into the inner block, the Button block is skipped and focus goes directly to the contenteditable.
Is there any focus management to skip the Button block? If so, I would consider to be consistent and manage focus to make it go directly to the social link button.

buttons

**Comment by getdave** [comment]
That's super helpful detail @afercia 👍

**Comment by jeryj** [comment]
> One option could be removing tabIndex from the blockProps. This way the LI, which is the actual inner block, wouldn't be focusable any longer and focus would go to the button. However, this would be an exception that is hard to justify as all blocks are supposed to be focusable.

That's what I tried initially, but it breaks using the delete button to delete a block.

> However, when tabbing into the inner block, the Button block is skipped and focus goes directly to the contenteditable.
Is there any focus management to skip the Button block? 

I believe if there is a focusable element inside the block, focus is supposed to be transferred to it automatically. Maybe there's a bug there that's breaking this. I'll investigate.

**Comment by jeryj** [comment]
> However, this would be an exception that is hard to justify as all blocks are supposed to be focusable.

Looked into this more. The current trunk behavior is technically correct if the `li` has the `blockProps`. We shouldn't automatically send focus to the `button`, as that would also remove the ability to delete an individual item or multiselect unless we also add these event listeners into the block. I also don't like the idea of duplicating behaviors in one block, as it will be a point to maintain a divergence. 

> However, when tabbing into the inner block, the Button block is skipped and focus goes directly to the contenteditable.

I think this is due to the button being a contenteditable. Also, as a result of this behavior, you can't have a selection on a single button or navigate between buttons easily. For example, To delete it, you have to delete all the text and press delete again. You can't arrow between the button blocks. I would prefer to be able to use the arrow keys to navigate between the button blocks instead of it automatically focusing the contenteditable.

https://github.com/user-attachments/assets/1e7b347d-e20b-4bfd-8530-1abf6b267bd1

> Navigating with Control+Option+Right arrow to the button, the button is not announced as button any longer. The only thin that is announced is the aria-label aria-label="Block: GitHub". No information that is'a an interactive button that opens a dialog.

If we preserve this route of moving the blockProps to the button, the main issue is that the social icon button block is not announced as a button so screen reader users don't have any idea that the element is interactive. If I can get the button to announce as a button instead of role=document, I think we could continue with the current route? Would that address the issue?

**Comment by jeryj** [comment]
I added the role=button as an override, and the semantics seem much better.

https://github.com/user-attachments/assets/16ed836e-d6f9-4be0-979a-48af9266ca13

**Comment by alexstine** [comment]
The block itself still really isn't focusable though. You think users will figure out they can press delete on the button to remove? The whole concept of blocks in Gutenberg drives me up a wall if I'm being honest. Block and block content, I have no idea how we ended up in this world where some have wrappers, some do not.

**Comment by jeryj** [comment]
> The block itself still really isn't focusable though.

Can you explain this more? It seems focusable in my testing. It's announcing "Block: Bandcamp, dialog pop up, button" and seems to have all the keybindings of a block.

**Review by alexstine — APPROVED** [review]
Gave this another review and it tests fine. Sorry for the really late reply.

**Comment by jeryj** [comment]
@afercia - Do you want to give this another round of testing before merging? I know it's a lot to go over.

**Review by getdave — APPROVED** [review]
I tested this and found the interactions to be much better. Combined with the changes to the inserter this whole block now feels a lot easier to use 🎉 

I tried to break this by changing colors, applying block styles and various things but I couldn't see any regressions.

Not sure if you wanted to add a quick set of tests here to ensure there are no future regressions to the focus handling?

Overall, great job 🙇

**Comment by afercia** [comment]
@jeryj thanks for the ping. Keyboard interaction is way better now, thank you!

> I added the role=button as an override, and the semantics seem much better.

Genuine question, I don't remember: why not 'omit' the `role` from `blockProps` instead of overriding it with an explicit `role="button"`? Something like: `const { role, ...blockPropsNoRole } = blockProps;`.

Nitpick:
at this point, the semantic structure would ne:
```
focusable `<ul>` element with role=document, semantically it's not an unordered list
    `<li>` elements are still list elements
        `<button>` element with role=button 
```

Basically the `<li>` elements are stray list items with no list parent. To be a little cleaner I'd suggest to add a `role="presentation"` to the list items.

**Comment by jeryj** [comment]
> Genuine question, I don't remember: why not 'omit' the role from blockProps instead of overriding it with an explicit role="button"? Something like: const { role, ...blockPropsNoRole } = blockProps;.

@afercia Thanks for reviewing this again! If we don't explicitly add the `button` role, `useBlockProps` adds a `role=document` to it, which removes the info about the popover attached to the `button`.

> Basically the `<li>` elements are stray list items with no list parent. To be a little cleaner I'd suggest to add a role="presentation" to the list items.

Thank you for that catch! Updated 👍🏻
